### PR TITLE
[RDY] Action parameter checking

### PR DESCRIPTION
**Pull request by TheCycoONE**
The primary purpose of this pr is to add type checking to the action queue and enforce consistent types for given properties to make errors easier to trace.

During the process I also found a typo which is corrected in the first commit.

### CODE REVIEW

File: CorsixTH/Lua/humanoid_actions/idle.lua [hunk 1]
```diff
@@ -33,6 +33,11 @@ end
 --!param direction (string) Direction of facing.
 --!return (action) Self, for daisy-chaining.
 function IdleAction:setDirection(direction)
+  assert(direction == nil or
+      direction == "north" or direction == "south" or
+      direction == "east" or direction == "west",
+      "direction must be north, south, east, or west")
```
Comment (Alberth289346) [inline CorsixTH/Lua/humanoid_actions/idle.lua]: Misses 'nil' as possible value

I am not so sure about the usefulness of the messages, as they seem to duplicate the code, but if you like them, why not.

File: CorsixTH/Lua/humanoid_actions/idle.lua [hunk 2]
```diff
@@ -41,6 +46,9 @@ end
 --!param on_interrupt (function) Function to call on interrupt.
 --!return (action) Self, for daisy-chaining.
 function IdleAction:setOnInterrupt(on_interrupt)
+  assert(on_interrupt == nil or type(on_interrupt) == "function",
+      "on_interrupt not a function")
```
Comment (Alberth289346) [inline CorsixTH/Lua/humanoid_actions/idle.lua]: missing nil as value

File: CorsixTH/Lua/humanoid_actions/knock_door.lua [hunk 3]
```diff
@@ -27,6 +27,11 @@ local KnockDoorAction = _G["KnockDoorAction"]
 --!param door (Object) Door to knock on.
 --!param direction (string) Direction of facing.
 function KnockDoorAction:KnockDoorAction(door, direction)
+  assert(class.is(door, Door), "expecting door object")
+  assert(direction == "north" or direction == "south" or
```
Comment (Alberth289346) [inline CorsixTH/Lua/humanoid_actions/knock_door.lua]: maybe refactor the direction test out into a separate function? it seems to happen a lot.

File: CorsixTH/Lua/humanoid_action.lua [hunk 4]
```diff
@@ -65,6 +74,9 @@ end
 --!param after_use (func) Callback function that is called after the action ends.
 --!return (action) Returning self, for daisy-chaining.
 function HumanoidAction:setAfterUse(after_use)
+  assert(after_use == nil or type(after_use) == "function",
+      "after_use must be a function")
```
Comment (Alberth289346) [inline CorsixTH/Lua/humanoid_action.lua]: misses nil in the error message

File: CorsixTH/Lua/humanoid_action.lua [hunk 5]
```diff
@@ -57,6 +63,9 @@ end
 --! termination conditions.
 --!return (action) Returning self, for daisy-chaining.
 function HumanoidAction:setLoopCallback(loop_callback)
+  assert(loop_callback == nil or type(loop_callback) == "function",
+      "loop_callback must be a function")
```
Comment (Alberth289346) [inline CorsixTH/Lua/humanoid_action.lua]: misses nil in the error message

**Comment by TheCycoONE** [comment]
I found the message helpful to know the parameter, otherwise you just get the function and line number in the error message. The actual parameter restriction is unnecessary though, so maybe I'll just change them all to 'Invalid value for x!'

File: CorsixTH/Lua/humanoid_actions/knock_door.lua (hunk 3 above)
Comment (TheCycoONE) [inline CorsixTH/Lua/humanoid_actions/knock_door.lua]: I considered it but I don't know where I would put it.

Ideally I have in mind something like class.lua but for simple enums, so we'd write something like ```enum('Direction', { 'north', 'south', 'east', 'west' })``` then everywhere we would write ```enum.is(point.direction, Direction)```

If you don't mind I won't do that for this pr.

**Comment by Alberth289346** [comment]
In C/C++ it also dumps the failed assert, I might be confused with that.

I agree a message is useful. I was mostly worried about the fact that you write the same thing both in code and in words. On the other hand, if you run into this thing in development, it's nice to  get a description of what is expected immediately without having to check those lines.

Both descriptions are fine to me, do what your :heart: tells you, or throw a :game_die: to decide.

**Comment by TheCycoONE** [comment]
Travis failure appears to be due to a change in Busted, unrelated to this PR.

I've updated the PR to change the assert message to ```Invalid value for parameter 'foo'``` everywhere.
I also created a ```HumanoidAction:afterLoad``` function and update the actions to conform to the proper parameter types there instead of `humanoid.lua`

**Comment by TheCycoONE** [comment]
Looks like Busted was fixed.

**Pull request merged**